### PR TITLE
For EVENT_PROCESSORs, derive the destinationType (queue|topic) from e…

### PR DESCRIPTION
**Pull request by arcadiushmcts**
…vent_sources.location.jms_uri in event-sources.yaml when RAML is used for the EVENT_PROCESSOR

### CODE REVIEW

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+ * This class is allowing to read any event-sources.yaml files in the classpath and trying to find the corresponding jms_uri
+ * that is used to determine whether we are dealing with a topic or a queue.
+ * Note that this is only used for event-processors only
+ */
+public class EventSourcesDestinationTypeFinder {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventSourcesDestinationTypeFinder.class);
+
+    private final JmsUriToDestinationTypeConverter uriToDestinationTypeConverter = new JmsUriToDestinationTypeConverter();
+    JmsUriToDestinationConverter jmsUriToDestinationConverter = new JmsUriToDestinationConverter();
```
Comment (allanmckenzie) [inline framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java:34]: Should JmsUriToDestinationConverter be private and final like the others?

File: framework-generators/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinderTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        final Optional<Class<? extends Destination>> res = destinationTypeFinder.findForEventProcessor(EVENT_PROCESSOR, "structure.event");
+
+        assertThat(res, Matchers.is(Optional.of(Topic.class)));
+    }
+
+    @Test
+    void shouldFindDestinationTypeForEventProcessorDefinedAsQueue() {
+        final Optional<Class<? extends Destination>> res = destinationTypeFinder.findForEventProcessor(EVENT_PROCESSOR, "queuestructure.event");
+
+        assertThat(res, Matchers.is(Optional.of(Queue.class)));
```
Comment (allanmckenzie) [inline framework-generators/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinderTest.java:31]: static import?

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+
+    /**
+     * @param serviceComponent EVENT_PROCESSOR, etc
+     * @param destinationName  the matching destination name
+     * @return Queue, Topic or empty
+     */
+    public Optional<Class<? extends Destination>> findForEventProcessor(final String serviceComponent, final String destinationName) {
+
+        if (serviceComponent == null || destinationName == null || !uriToDestinationTypeConverter.isEventProcessor(serviceComponent)) {
```
Comment (allanmckenzie) [inline framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java:45]: This is complex logic, and would need three tests just to test the above line. I'm assuming it's tested, though it's hard to be sure. Maybe extract this logic to another class to make testing easier?

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            return empty();
+        }
+
+        try {
+            final EventSourcesFileParser eventSourcesFileParser = eventSourceYamlParserFactory.create();
+            final ArrayList<Path> list = new ArrayList<>();
+            final List<EventSourceDefinition> eventSourceDefinitionList = eventSourcesFileParser.getEventSourceDefinitions(Path.of(""), list);
+            if (eventSourceDefinitionList == null) {
+                LOGGER.debug("No SubscriptionWrapper found .. returning");
```
Comment (allanmckenzie) [inline framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java:54]: Do we need these debug log messages? I'm guessing you added them to make sure stuff is working, but now that it is, are they still needed? If they are then the logger should be converted to a member variable (i.e. not static), mocked in the tests and verified

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            final Optional<Class<? extends Destination>> finalDestination = destinations.stream().
+                    findFirst().
+                    get();
+
+            LOGGER.debug("FinalDestination is {}", finalDestination);
+
+            return finalDestination;
+
+        } catch (Exception e) {
+            LOGGER.warn("Failed in findForEventProcessor: {}", e.getMessage());
```
Comment (allanmckenzie) [inline framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java:95]: This log message should definitely be tested

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            return empty();
+        }
+
+    }
+
+    private String uriToDestination(final String jmsUri) {
+        try {
+            return jmsUriToDestinationConverter.convert(jmsUri);
+        } catch (Exception e) {
+            return "";
```
Comment (allanmckenzie) [inline framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java:105]: Is an empty String the best thing to return if the uri conversion fails?

**Review by allanmckenzie — APPROVED** [review]

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java (hunk 1 above)
Comment (arcadiushmcts) [reply framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java:34]: Fixed.
 I was just testing your vigilence :smile: @allanmckenzie

File: framework-generators/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinderTest.java (hunk 2 above)
Comment (arcadiushmcts) [reply framework-generators/messaging-adapter-generator/src/test/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinderTest.java:31]: Fixed

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java (hunk 4 above)
Comment (arcadiushmcts) [reply framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java:54]: I have removed those debug statements

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java (hunk 3 above)
Comment (arcadiushmcts) [reply framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java:45]: actually, only 2 conditions are needed:
``` if (destinationName == null || !uriToDestinationTypeConverter.isEventProcessor(serviceComponent))```

We have only 1 test 
```
 @Test
    void shouldFindDestinationTypeForEventProcessorAsEmptyWhenComponentIsNull() {
        final Optional<Class<? extends Destination>> res = destinationTypeFinder.findForEventProcessor(null,null);

        assertThat(res, is(empty()));
    }
```

I have added more tests....  there are now 5 of them

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java (hunk 6 above)
Comment (arcadiushmcts) [reply framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java:105]: This utility is used in stream.filer(something.equalsIgnoreCase(thisMethod(uri))) 
We could return optional... returning empty string here simplifies the code.

File: framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java (hunk 5 above)
Comment (arcadiushmcts) [reply framework-generators/messaging-adapter-generator/src/main/java/uk/gov/justice/raml/jms/core/EventSourcesDestinationTypeFinder.java]: test added

**Review by allanmckenzie — APPROVED** [review]

**Pull request merged**